File: backend/src/Raven.Api/Features/Profiles/ProfileReadiness.cs
using Raven.Api.Features.Research.Coverage;

namespace Raven.Api.Features.Profiles;

/// <summary>
/// Describes how much supported research a profile contains. A profile row is
/// not automatically a usable profile: a failed or stale workflow must not be
/// able to turn an identity-only snapshot into a Profile Improvement baseline.
/// </summary>
public enum CompanyProfileReadinessState
{
    IdentityOnly,
    Sparse,
    Partial,
    Complete
}

public static class CompanyProfileReadiness
{
    public const int RequiredTargetCount = 9;
    public const int PartialMinimumCoveredTargets = 3;

    public static CompanyProfileReadinessState GetState(CompanyProfileSnapshot profile)
    {
        ArgumentNullException.ThrowIfNull(profile);
        var covered = CountCoveredTargets(profile);
        return covered switch
        {
            0 => CompanyProfileReadinessState.IdentityOnly,
            >= RequiredTargetCount => CompanyProfileReadinessState.Complete,
            >= PartialMinimumCoveredTargets => CompanyProfileReadinessState.Partial,
            _ => CompanyProfileReadinessState.Sparse
        };
    }

    public static int CountCoveredTargets(CompanyProfileSnapshot profile)
    {
        ArgumentNullException.ThrowIfNull(profile);
        var count = 0;
        if (!string.IsNullOrWhiteSpace(profile.LegalName)) count++;
        if (!string.IsNullOrWhiteSpace(profile.RegistrationNumberOrTaxId)) count++;
        if (profile.FoundedYear.HasValue) count++;
        if (!string.IsNullOrWhiteSpace(profile.PrimaryIndustry) || profile.SecondaryIndustries.Count > 0) count++;
        if (!string.IsNullOrWhiteSpace(profile.CompanySize)
            || profile.EmployeeCount.HasValue
            || !string.IsNullOrWhiteSpace(profile.EmployeeCountRange)) count++;
        if (profile.ProductsServices.Count > 0) count++;
        if (profile.Markets.Count > 0) count++;
        if (profile.Leadership.Count > 0) count++;
        if (profile.Locations.Count > 0) count++;
        return count;
    }

    /// <summary>
    /// The generated profile provenance is deliberately separate from a
    /// company name. It lets the server distinguish a confirmed model result
    /// from an accidental identity-only row left by an interrupted workflow.
    /// </summary>
    public static bool HasModelCreationProvenance(CompanyProfileSnapshot profile) =>
        profile switch
        {
            CompanyProfileCandidate candidate => HasModelCreationProvenance(candidate.AiProvider, candidate.AiModel, candidate.PromptTemplateVersion),
            CompanyProfileVersion version => HasModelCreationProvenance(version.AiProvider, version.AiModel, version.PromptTemplateVersion),
            _ => false
        };

    public static bool IsUsableAcceptedProfile(CompanyProfileVersion profile) =>
        HasModelCreationProvenance(profile) && CountCoveredTargets(profile) > 0;

    public static bool IsConfirmableCandidate(CompanyProfileCandidate candidate) =>
        HasModelCreationProvenance(candidate)
        && CountCoveredTargets(candidate) > 0
        && candidate.Evidence.Any(evidence => evidence.SourceDocumentIds.Count > 0);

    private static bool HasModelCreationProvenance(string? provider, string? model, string? promptTemplateVersion) =>
        !string.IsNullOrWhiteSpace(provider)
        && !string.IsNullOrWhiteSpace(model)
        && !string.IsNullOrWhiteSpace(promptTemplateVersion);
}

File: frontend/src/utils/profileReadiness.ts
export interface ProfileReadinessSource {
  aiProvider?: string | null;
  aiModel?: string | null;
  promptTemplateVersion?: string | null;
  legalName?: string | null;
  registrationNumberOrTaxId?: string | null;
  foundedYear?: number | null;
  primaryIndustry?: string | null;
  secondaryIndustries?: string[] | null;
  companySize?: string | null;
  employeeCount?: number | null;
  employeeCountRange?: string | null;
  productsServices?: unknown[] | null;
  markets?: unknown[] | null;
  leadership?: unknown[] | null;
  locations?: unknown[] | null;
}

export type DossierProfileReadiness = "missing" | "identity-only" | "sparse" | "partial" | "complete";

const requiredTargetCount = 9;
const partialMinimumCoveredTargets = 3;

function hasModelCreationProvenance(profile: ProfileReadinessSource) {
  return Boolean(profile.aiProvider?.trim() && profile.aiModel?.trim() && profile.promptTemplateVersion?.trim());
}

export function countProfileCoveredTargets(profile: ProfileReadinessSource) {
  let covered = 0;
  if (profile.legalName?.trim()) covered += 1;
  if (profile.registrationNumberOrTaxId?.trim()) covered += 1;
  if (profile.foundedYear !== null && profile.foundedYear !== undefined) covered += 1;
  if (profile.primaryIndustry?.trim() || (profile.secondaryIndustries?.length ?? 0) > 0) covered += 1;
  if (profile.companySize?.trim() || (profile.employeeCount !== null && profile.employeeCount !== undefined) || profile.employeeCountRange?.trim()) covered += 1;
  if ((profile.productsServices?.length ?? 0) > 0) covered += 1;
  if ((profile.markets?.length ?? 0) > 0) covered += 1;
  if ((profile.leadership?.length ?? 0) > 0) covered += 1;
  if ((profile.locations?.length ?? 0) > 0) covered += 1;
  return covered;
}

export function getProfileReadiness(profile?: ProfileReadinessSource | null): DossierProfileReadiness {
  if (!profile) return "missing";
  const covered = countProfileCoveredTargets(profile);
  if (covered === 0) return "identity-only";
  if (covered >= requiredTargetCount) return "complete";
  if (covered >= partialMinimumCoveredTargets) return "partial";
  return "sparse";
}

/**
 * A version must be a model-created, evidence-bearing profile with at least
 * one supported research target before it can be used as a patch baseline.
 * The identity-only row left by an interrupted run is intentionally excluded.
 */
export function hasUsableAcceptedProfile(profile?: ProfileReadinessSource | null, gapCount?: number) {
  return Boolean(
    profile
    && hasModelCreationProvenance(profile)
    && countProfileCoveredTargets(profile) > 0
    && (gapCount === undefined || gapCount < requiredTargetCount),
  );
}

export function profileHasModelCreationProvenance(profile?: ProfileReadinessSource | null) {
  return Boolean(profile && hasModelCreationProvenance(profile));
}
